File: AlgorithmDataStructure/LeetCode/695-MaxAreaOfIsland/max.swift

class Solution {
    
    //广度优先搜索
    //遍历网格，是1的进入队列，然后遍历队列，上下左右计算面积
    func maxAreaOfIsland(_ grid: [[Int]]) -> Int {
        var ans = 0, tempGrid = grid
        let dirX = [0,1,0,-1], dirY = [1,0,-1,0]
        
        for i in 0 ..< tempGrid.count {
            for j in 0 ..< tempGrid[i].count {
                if tempGrid[i][j] != 1 { continue }
                var count = 0
                var dirXY = [(Int,Int)]()
                dirXY.append((i, j))
                
                while dirXY.isEmpty == false {
                    let tempXY = dirXY.removeFirst()
                    let tempX = tempXY.0, tempY = tempXY.1
                    if tempX >= 0 && tempX < tempGrid.count && tempY >= 0 && tempY < tempGrid[i].count && tempGrid[tempX][tempY] == 1 {
                        count += 1
                        tempGrid[tempX][tempY] = 0
                    } else {
                        continue
                    }
                    
                    for index in 0 ..< 4 {
                        dirXY.append((tempX + dirX[index], tempY + dirY[index]))
                    }
                }
                ans = max(ans, count)
            }
            
        }
        
        return ans
    }
    
}

let s = Solution()
print(s.maxAreaOfIsland([[0,1,1],[0,1,0],[0,0,0]]))

File: AlgorithmDataStructure/LeetCode/695-MaxAreaOfIsland/max.c

#include <stdio.h>
#include <stdlib.h>

//深度优先搜索
//遇到是1的点就，遍历上下左右，计算过的点置为0
int dfs(int** grid, int gridSize, int* gridColSize, int x, int y) {
    if (x < 0 || y < 0 || x >= gridSize || y >= gridColSize[0] || grid[x][y] == 0) {
        return 0;
    }
    grid[x][y] = 0; //计算过了就置0
    int dirX[] = {0,1,0,-1};
    int dirY[] = {1,0,-1,0};
    int ans = 1;
    for (int i = 0; i < 4; i++) {
        int nextX = x + dirX[i], nextY = y + dirY[i];
        ans += dfs(grid, gridSize, gridColSize, nextX, nextY);
    }
    return ans;
}


int maxAreaOfIsland(int** grid, int gridSize, int* gridColSize){
    if (grid == NULL || gridSize <= 0 || gridColSize == NULL ) { return 0; }
    
    int ans = 0;
    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridColSize[0]; j++) {
            if (grid[i][j] == 1) {
                int temp = dfs(grid, gridSize, gridColSize, i, j);
                ans = ans < temp ? temp : ans;
            }
        }
    }
    return ans;
}


int main(int argc, const char * argv[]) {
    
    int array[3][3] = {{0,1,0},{1,1,0},{0,0,1}};
    
    int **a = (int **)malloc(sizeof(int *)*3);
    for (int i = 0; i < 3; i++) {
        int *a1 = (int *)malloc(sizeof(int)*3);
        for (int j = 0; j < 3; j++) {
            a1[j] = array[i][j];
        }
        a[i] = a1;
    }
    int b[3] = {3};
    
    printf("%d \n", maxAreaOfIsland(a, 3, b));
    
    return 0;
}




